[#167] Unnecessary catch block in MainApp and Exceptions in Interfaces

Eclipse-Dominator opened the pull request:
Fixes #167 

Alternative Approach: #191 

Classes that implements the methods,
- readAddressBook() from AddressBookStorage interface
- readUserPrefs() from UserPrefsStorage interface does not actually throw any IOException as such exceptions are caught within the method and is being treated as DataConversionException which may not be the case.

All IOException are caught by JsonUtil.readJsonFile() which is called by classes implementing these methods.

This method also causes ConfigUtil.readConfig() to not throw IOExceptions when handling actual IOExceptions that is not caused by a conversion error.

This causes the wrong log messages to be logged when an actual IOException occur.

Let's
- decouple actual conversion errors from IOExceptions and restore functionality to the catch blocks in MainApp.java that handles IOExceptions.
- add catch blocks to handle IOException when reading config files

---

We can see the effect of the change below when reading files that does not have sufficient read permissions:

![image](https://github.com/se-edu/addressbook-level3/assets/4567895/19810d22-00db-4e70-bc9c-5f4eba137791)

---

Right now, actual IOExceptions (not conversion errors) are still being handled like conversion errors where a blank/default save file is used instead. 
I would prefer to show some form of warning then exit the application instead. However, this might require additional changes to UI behavior so I left it as such for now.

damithc commented on src/main/java/seedu/address/MainApp.java:
```diff
@@ -128,6 +128,10 @@ protected Config initConfig(Path configFilePath) {
             initializedConfig = new Config();
+        } catch (IOException e) {
+            logger.warning("Problem while reading from the file " + configFilePathUsed
+                    + ". Will be starting with default configuration");
```
Why is this part different from the `DataConversionException` catch clause?

Eclipse-Dominator replied:
I transferred from the other file checks since Config initially did not have the catch block for IOException. I will standardize the phrasing.

damithc commented on src/main/java/seedu/address/MainApp.java:
```diff
@@ -128,6 +128,10 @@ protected Config initConfig(Path configFilePath) {
+        } catch (IOException e) {
+            logger.warning("Problem while reading from the file " + configFilePathUsed
+                    + ". Will be starting with default configuration");
+            initializedConfig = new Config();
```
Is it possible to combine these two catch clauses, as the action taken is the same?  

```java
        } catch (DataConversionException | IOException e) {
            logger.warning(e.getMessage + " Using default config properties.");
            initializedConfig = new Config();
        }
```

Eclipse-Dominator replied:
`DataConversionException` is a custom exception that actually takes in a file path. So when `e.getMessage()` is called on it, it returns a file path instead of the exception as compared to `IOException`.

damithc replied:
> `DataConversionException` is a custom exception that actually takes in a file path.

I don't think so. Check the code.

Eclipse-Dominator replied:
I got confused with some other repo. 

I have checked the `getMessage` when a DataConversionException is created, it seems that such exceptions have a really long message:

![image](https://github.com/se-edu/addressbook-level3/assets/4567895/fe33013d-26e2-48b7-8428-7aaac9136686)

The warning is actually already logged prior to this by `readJsonFile` 
![image](https://github.com/se-edu/addressbook-level3/assets/4567895/504dc1d4-d0a6-4e2f-91a7-6898ff98758b)

so it might better if we just provide a short description of what happened instead?

Alternatively, we can specifically override the `getMessage()` for `DataConversionException` to give a description of the error since the details are already printed prior to this?

---

Or we can just change all the catch block to say error reading from file since the reason is already logged by `readJsonFile`

damithc replied:
>so it might better if we just provide a short description of what happened instead?

How would the code look like for that option?

Eclipse-Dominator replied:
DataConversionException: "XXX file at <file location> is might be corrupted/in the wrong format\n"
IOException: "An error occurred while reading from XXX file at <file location>: <e.getMessage()>\n"

We can perhaps combine them into a singular catch block like so:

```java
        } catch (DataConversionException | IOException e) {
            String description = "";
            if (e instanceof DataConversionException) {
                description = "Config file at " + configFilePathUsed
                        + " is might be corrupted or in the wrong format";
            } else {
                description = "Problem while reading from the file " + configFilePathUsed;
            }
            logger.warning(description + ". Using default config properties.");
            initializedConfig = new Config();
        }
```

The result will look like so:
![image](https://github.com/se-edu/addressbook-level3/assets/4567895/a302ac84-bdb9-41f2-b7e6-1ea86ea54733)

damithc replied:
`if (e instanceof DataConversionException)` defeats the purpose of combining the two `catch` clauses. In any case, I just noticed that the `initModelManager` method uses two `catch` clauses as well. So, we can do the same here.

damithc reviewed (changes requested):
Some initial comments ...

damithc reviewed (commented):
@Eclipse-Dominator you can proceed to the next iteration of this PR. 👍

Eclipse-Dominator committed:
Restore functionality to unreachable catch blocks

Classes that implements the methods,
- readAddressBook() from AddressBookStorage interface
- readUserPrefs() from UserPrefsStorage interface
does not actually throw any IOException as such exceptions are caught
within the method and is being treated as DataConversionException
which may not be the case.

All IOException are caught by JsonUtil.readJsonFile() which is called by
classes implementing these methods.

This method also causes ConfigUtil.readConfig() to not throw
IOExceptions when handling actual IOExceptions that is not caused by
a conversion error.

This causes the wrong log messages to be logged when an actual
IOException occur.

Let's
- decouple actual conversion errors from IOExceptions and restore
functionality to the catch blocks in MainApp.java that handles
IOExceptions.
- add catch blocks to handle IOException when reading config files
- reduce code duplication in catch blocks

Eclipse-Dominator commented:
I have included some change to reduce code duplication in handling of config and pref files.

damithc commented on src/main/java/seedu/address/commons/util/JsonUtil.java:
```diff
@@ -68,9 +68,12 @@ public static <T> Optional<T> readJsonFile(
         try {
             jsonFile = deserializeObjectFromJsonFile(filePath, classOfObjectToDeserialize);
+        } catch (JsonProcessingException e) {
+            logger.warning("Error processing data from jsonFile file " + filePath + ": " + e);
```
```suggestion
            logger.warning("Error processing data from JSON file " + filePath + ": " + e);
```

damithc reviewed (commented):
Pointed a minor nit.

Good job writing a detailed commit message. I think the key statement is the following.
>This causes the wrong log messages to be logged when an actual
>IOException occur.

We should lead with that, and elaborate more. What kind of wrong message is logged in what kind of situuation? So, the commit message can start like this:

> When X happens, the incorrect message Y is logged.
>
> Briefly describe why the above happens.
>
> Briefly describe how this PR fixes it.

damithc commented:
@se-edu/tech-team-level1 for your reviews please ...

mayuanxin1234 commented on src/main/java/seedu/address/commons/util/JsonUtil.java:
```diff
@@ -68,9 +68,12 @@ public static <T> Optional<T> readJsonFile(
+            logger.warning("Error processing data from jsonFile file " + filePath + ": " + e);
+            throw new DataConversionException(e);
         } catch (IOException e) {
             logger.warning("Error reading from jsonFile file " + filePath + ": " + e);
```
```suggestion
            logger.warning("Error reading from JSON file " + filePath + ": " + e);
```

SPWwj commented on src/main/java/seedu/address/commons/util/JsonUtil.java:
```diff
@@ -68,9 +68,12 @@ public static <T> Optional<T> readJsonFile(
         } catch (IOException e) {
             logger.warning("Error reading from jsonFile file " + filePath + ": " + e);
-            throw new DataConversionException(e);
+            throw e;
```
I've noticed a potential issue with the error handling strategy in this code. Specifically, the practice of catching and immediately re-throwing the same error could lead to some inefficiencies.

In your current setup, you've employed a logger to document exceptions. However, by re-throwing the same exception, it's possible that another component, such as MainApp, might process this exception once more. This could inadvertently lead to duplicate entries in our logs, complicating the debugging process and adding unnecessary clutter to our log files.

To enhance this, I'd suggest revisiting the error handling strategy.

damithc replied:
Good observations @SPWwj 
I wouldn't worry about the efficiency so much though. The difference in performance is unlikely to matter.
There is value is logging an exception along multiple points between where it was thrown and where it was handled. For one thing, a similar exception can be thrown from multiple places but handled in one place, which means it is useful to record where it was thrown. For another, the decision to log or not is best made locally (i.e., without considering where the method is being used) -- as the usage of a method should be able to change freely without having to change the method. So, we shouldn't make any assumptions about how the method will be used and how the exception will be handled over there.

SPWwj reviewed (commented):
Upon reviewing this Pull Request (PR), it appears that it may not adequately address the problem outlined in issue #167. Specifically, the changes seem to make the IOException catch block mandatory in MainApp, which may not be the intended solution.

I would recommend considering a different approach to resolve this issue. If this PR is addressing a different problem, it might be more appropriate to create a new issue to track and discuss these changes separately. This will help maintain clarity and focus in our issue tracking.


<img width="1800" alt="image" src="https://github.com/se-edu/addressbook-level3/assets/30100720/b5913502-99a4-4595-bbd0-000ba41e4bdd">
I feel the issue at hand is more related to the appropriateness of the error message rather than incorrect error handling.

chia-yh commented on src/main/java/seedu/address/MainApp.java:
```diff
@@ -84,11 +84,12 @@ private Model initModelManager(Storage storage, ReadOnlyUserPrefs userPrefs) {
         } catch (DataConversionException e) {
-            logger.warning("Data file at " + storage.getAddressBookFilePath() + " is not in the correct format."
+            logger.warning("Data file at " + storage.getAddressBookFilePath()
+                    + " is might be corrupted or in the wrong format."
```
```suggestion
                    + " might be corrupted or in the wrong format."
```

jundatan commented on src/main/java/seedu/address/MainApp.java:
```diff
@@ -106,30 +107,31 @@ private void initLogging(Config config) {
+                    + " might be corrupted or in the wrong format"
+                    + ". Using default config properties.");
+        } catch (IOException e) {
+            logger.warning("Problem while reading from the config file " + configFilePathUsed
```
```suggestion
            logger.warning("Problem while reading from the config file at " + configFilePathUsed
```

jundatan reviewed (commented):
Suggested change for warning logger.

damithc commented:
>Upon reviewing this Pull Request (PR), it appears that it may not adequately address the problem outlined in issue https://github.com/se-edu/addressbook-level3/issues/167. Specifically, the changes seem to make the IOException catch block mandatory in MainApp, which may not be the intended solution.

I agree what what @SPWwj  said above in that there is disparity between the issue and the PR. The original intention was to remove the unnecessary catch block.  I prefer that approach as it simplifies the code without any loss of value (because we do not currently have different handlings for incorrect format in JSON files vs other I/O problems with the JSON files). 

In addition, The name `DataConversionException` may be misleading because we use it to represent both kind of problems mentioned above. Perhaps something like `ConfigLoadingException` is more representative?

SPWwj commented:
> `DataConversionException`

The renaming of the exception to ConfigLoadingException aligns well with its purpose.

In JsonUtil.java, handling the IOException by logging it provides valuable information about the location of the error, aiding in debugging.

In MainApp, handling the ConfigLoadingException manages any issues that arise during the data fetching process, enhancing the robustness of error handling and preventing application crashes due to unhandled exceptions.

The use of interfaces for the readUserPrefs method allows for various implementations of the logic. It's not restricted to reading data from a local JSON file; it could also fetch data over the internet. As a result, different types of exceptions, not just IOException, could be encountered. By using ConfigLoadingException as a general exception, these scenarios can be handled in a unified way.

linustws commented on src/main/java/seedu/address/MainApp.java:
```diff
@@ -148,22 +150,20 @@ protected UserPrefs initPrefs(UserPrefsStorage storage) {
         Path prefsFilePath = storage.getUserPrefsFilePath();
         logger.info("Using preference file : " + prefsFilePath);
```
```suggestion
        logger.info("Using preference file at " + prefsFilePath);
```

Eclipse-Dominator commented:
> > Upon reviewing this Pull Request (PR), it appears that it may not adequately address the problem outlined in issue #167. Specifically, the changes seem to make the IOException catch block mandatory in MainApp, which may not be the intended solution.
> 
> I agree what what @SPWwj said above in that there is disparity between the issue and the PR. The original intention was to remove the unnecessary catch block. I prefer that approach as it simplifies the code without any loss of value (because we do not currently have different handlings for incorrect format in JSON files vs other I/O problems with the JSON files).
> 
> In addition, The name `DataConversionException` may be misleading because we use it to represent both kind of problems mentioned above. Perhaps something like `ConfigLoadingException` is more representative?

Hmm, it's actually possible to remove the unnecessary catch block as indicated in the original PR as technically IO related errors/logs can be handled within the `jsonutil` method and we only need to mention that something went wrong while trying to read XXX file in the `MainApp`.

My original concern is that students may want to implement additional features/error handling when encountering these errors. Thus, I went ahead with restoring the functionality of the IOException catch block. 

As this is also another possibility, I will create a separate PR removing the catch blocks as intended in the original issue to explore other approaches to this. You can find the alternative approach in #191

damithc commented:
Closing this in favor of #191
